smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -532,20 +569,326 @@ private void generateOperationDeserializer(
+                    });
+                }
+            });
+        }
+    }
+
+    private void readResponseBody(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        List<HttpBinding> documentBindings = bindingIndex.getResponseBindings(operation, Location.DOCUMENT);
+        documentBindings.sort(Comparator.comparing(HttpBinding::getMemberName));
+        List<HttpBinding> payloadBindings = bindingIndex.getResponseBindings(operation, Location.PAYLOAD);
+
+        if (!documentBindings.isEmpty()) {
+            deserializeDocument(context, operation, documentBindings);
+        }
+        if (!payloadBindings.isEmpty()) {
+            // There can only be one payload binding.
+            HttpBinding binding = payloadBindings.get(0);
+            Shape target = context.getModel().getShapeIndex().getShape(binding.getMember().getTarget()).get();
+            writer.write("output.$L = $L;", binding.getMemberName(), getOutputValue(
+                    context, Location.PAYLOAD, "data", operation, target));
+        }
+    }
+
+    private void readErrorBody(GenerationContext context, StructureShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+        String methodName = ProtocolGenerator.getDeserFunctionName(symbol, getName());
+        String contentName = StringUtils.uncapitalize(shape.getId().getName());
+
+        writer.openBlock("const $L = (\n"
+                       + "  output: any,\n"
+                       + "  context: SerdeContext\n"
+                       + "): $T => {", "};", methodName, symbol, () -> {
+            // Initial contents of the error.
+            writer.openBlock("let $L: $T = {", "};", contentName, symbol, () -> {
+                writer.write("__type: $S,", shape.getId().toString());
+                writer.write("$$name: $S,", shape.getId().getName());
+                writer.write("$$fault: $S,", shape.getTrait(ErrorTrait.class).get().getValue());
+            });
+            // Read additional modeled content.
+            deserializeError(context, shape);
+            writer.write("return $L;", contentName);
+        });
         writer.write("");
     }
+
+    protected String getOutputValue(
+            GenerationContext context,
+            Location bindingType,
+            String dataSource,
+            Shape container,
+            Shape target
+    ) {
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        if (container instanceof UnionShape && !unionTargetNeedsSerde(context.getModel().getShapeIndex(), target)) {
+            return dataSource;
+        }
+
+        if (target instanceof StringShape || target instanceof DocumentShape
```

Comment:
Consider also extracting this into a method with a clear name for readability

Maybe we need to add methods to Shape like isSimpleType?